### PR TITLE
Fix a bug in AsyncAllocator memory pools access setting.

### PR DESCRIPTION
**Pull request by buptzyb**
The original code uses the wrong parameter `i` for `cuDeviceCanAccessPeer()`, which can cause undefined behavior when `(*all_ids_)[i] != platform_device_id`, for example, virtual devices are used. Should use `(*all_ids_)[i].value()` to represent the previous pool id.

Also, add a judgment to skip the pool initialization process if it's already initialized previously. This can happen in some cases, such as when multiple virtual devices are created from one physical GPU, the virtual devices will actually share the same CUDA memory pool.

### CODE REVIEW

File: third_party/xla/xla/stream_executor/gpu/gpu_cudamallocasync_allocator.cc [hunk 1]
```diff
@@ -215,6 +215,15 @@ GpuCudaMallocAsyncAllocator::GpuCudaMallocAsyncAllocator(
   static auto* all_pools_ = new std::vector<CUmemoryPool*>();
   static auto* all_ids_ = new std::vector<tsl::PlatformDeviceId>();
   DCHECK(all_pools_->size() == all_ids_->size());
+  for (auto& pool_item_ : *all_pools_) {
+    if (*pool_item_ == pool_) {
```
Comment (changhuilin) [inline third_party/xla/xla/stream_executor/gpu/gpu_cudamallocasync_allocator.cc:219]: Just to check my understanding - each device has a default memory pool. Virtual devices on the same physical device will use the same memory pool. So for all virtual devices on the same physical device, we only initialize the pool of this physical device once. In other words, if the `PlatformDeviceId` already appears in the `all_ids_`, the pool initialization can be skipped. Right?

Can you add some comment for this?

File: third_party/xla/xla/stream_executor/gpu/gpu_cudamallocasync_allocator.cc [hunk 2]
```diff
@@ -246,10 +255,11 @@ GpuCudaMallocAsyncAllocator::GpuCudaMallocAsyncAllocator(
     // Set the previous pools access to the current GPU.
     map.location.id = platform_device_id.value();
 
-    VLOG(2) << "Set access to the pool id: " << i
+    VLOG(2) << "Set access to the pool id: " << (*all_ids_)[i].value()
```
Comment (changhuilin) [inline third_party/xla/xla/stream_executor/gpu/gpu_cudamallocasync_allocator.cc:258]: use a temp var to save `(*all_ids_)[i].value()`, and use it below?

File: third_party/xla/xla/stream_executor/gpu/gpu_cudamallocasync_allocator.cc (hunk 1 above)
Comment (buptzyb) [reply third_party/xla/xla/stream_executor/gpu/gpu_cudamallocasync_allocator.cc:226]: Yes, your understanding is correct. Add comments ahead.

File: third_party/xla/xla/stream_executor/gpu/gpu_cudamallocasync_allocator.cc (hunk 2 above)
Comment (buptzyb) [reply third_party/xla/xla/stream_executor/gpu/gpu_cudamallocasync_allocator.cc]: create a var `int previous_pool_id = (*all_ids_)[i].value();`.

**Review by changhuilin — APPROVED** [review]

**Comment by cheshire** [comment]
BTW please send XLA PRs via github.com/openxla/xla, it backs the same files, and soon will be the only way to submit such PRs.

**Comment by buptzyb** [comment]
> BTW please send XLA PRs via github.com/openxla/xla, it backs the same files, and soon will be the only way to submit such PRs.

XLA PR here: https://github.com/openxla/xla/pull/13106. @cheshire

**Comment by gbaned** [comment]
Hi @buptzyb Can you please resolve the conflicts? Thank you!

**Comment by keerthanakadiri** [comment]
Hi @buptzyb Can you please resolve the conflicts? Thank you!

**Comment by buptzyb** [comment]
Close this PR as it has been merged in XLA repo.

**Pull request closed without merging**